src/pages/ServiceDetails/ServiceDetailsPage.tsx

```diff
@@ -322,18 +323,29 @@ class ServiceDetails extends React.Component<ServiceDetailsProps, ServiceDetails
           }
         });
     } else {
-      const params = getQueryJaeger();
+      let params = changeParams(this.traceParams);
+      if (params) {
+        this.traceParams = params;
+        this.lastFetchTraceMicros = undefined;
+      } else {
+        params = getQueryJaeger();
+      }
+      if (this.lastFetchTraceMicros) {
+        params[URLParam.JAEGER_START_TIME] = this.lastFetchTraceMicros;
+      }
       fetchTraces(this.props.match.params.namespace, this.props.match.params.service, params)
         .then(traces => {
           const appendTraces = this.lastFetchTraceMicros !== undefined;
           this.lastFetchTracesError = false;
           let myState = {};
           if (traces && traces.data) {
             this.traces = appendTraces
-              ? this.traces.filter(t => t.startTime >= params[URLParam.JAEGER_START_TIME]).concat(traces.data)
+              ? this.traces.filter(t => t.startTime <= params[URLParam.JAEGER_START_TIME]).concat(traces.data)
```

Comment:
ok, this is the faulty check. It should be `>=` , but `params[URLParam.JAEGER_START_TIME]` should be here the start time in kiali time frame, not the one from the jaeger query (that's why I was mentioned doing a copy of params before changing it)

Reply:
Sorry I agree with the copy now I understand.

But I can't understand the second step.

If I have the traces from (t-1) to t, and I perform the refresh to request traces from t to t+1.

I should filter traces to have all the traces <= t and concatenate them with the new traces than I am requesting from t to  t+1.

So I can't udnerstand why in https://github.com/kiali/kiali-ui/blob/master/src/components/Metrics/SpanOverlay.ts#L10-L38
(I add comments in code)
```ts
const nowMicros = new Date().getTime() * 1000;
    const frameStart = nowMicros - (duration + 60) * 1000000; // seconds to micros with 1min margin;
// frameStart is my params[URLParam.JAEGER_START_TIME]
// calculated in RouterHelper
    const opts: TracingQuery = { startMicros: this.lastFetchMicros || frameStart };
    API.getServiceSpans(namespace, service, opts)
      .then(res => {
        this.lastFetchError = false;
// Here you have in this.spans spans from t-1 to frameStart
// so why are you filtering by spans later than frameStart ? this should be zero
        this.spans = doAppend ? this.spans.filter(s => s.startTime >= frameStart).concat(res.data) : res.data;
        this.onChange(this.buildOverlay());
        // Update last fetch time only if we had some results
        // So that if Jaeger DB hadn't time to ingest data, it's still going to be fetched next time
        if (this.spans.length > 0) {
          this.lastFetchMicros = Math.max(...this.spans.map(s => s.startTime));
        }
      })
      .catch(err => {
        if (!this.lastFetchError) {
          AlertUtils.addError('Could not fetch spans.', err);
          this.lastFetchError = true;
        }
      });
```
Maybe I need to grab another coffee

Reply:
`// Here you have in this.spans spans from t-1 to frameStart`

This isn't exactly correct. We must make the distinction between the frame start in Kiali (ie. time for whole graph) versus frame start for fetching data. Let's call them `kialiFrameStart` and `fetchFrameStart`.

So here, `this.spans` are spans before `fetchFrameStart`, and most of them are after `kialiFrameStart`. But because time flies, some of them can possibly be before `kialiFrameStart` and these are the ones we need to filter out, to keep only spans that fits in kiali time frame.